### PR TITLE
Introducing the usage of `HypervisorUpgradePlanner` in planning stage.

### PR DESCRIPTION
**Pull request by rgildein**
Introduce base structure of the `cou.steps.plan.generate_plan` generating the upgrade plan for data plane applications, right now it will generate only plan using `HypervisorUpgradePlanner` for all application deployed on same machine as Nova-compute.

### CODE REVIEW

File: cou/steps/plan.py [hunk 1]
```diff
@@ -317,6 +318,75 @@ def _get_os_release_and_series(analysis_result: Analysis) -> tuple[OpenStackRele
...
+
+    subordinate_upgrade_plan = create_upgrade_group(
+        apps=apps,
+        description="Control Plane subordinate(s) upgrade plan",
+        target=target,
+        force=force,
+        filter_function=lambda app: isinstance(app, SubordinateBaseClass),
+    )
+
+    logger.debug("generating of control plan upgrade plan finished")
```
Comment (gabrielcocenza) [inline cou/steps/plan.py:351]:
```suggestion
    logger.debug("generating of control plan upgrade plane finished")
```

File: cou/steps/plan.py [hunk 2]
```diff
@@ -354,28 +424,16 @@ async def generate_plan(analysis_result: Analysis, args: CLIargs) -> UpgradePlan
...
-        apps=analysis_result.apps_control_plane,
-        description="Control Plane subordinate(s) upgrade plan",
-        target=target,
-        force=args.force,
-        filter_function=lambda app: isinstance(app, SubordinateBaseClass),
+    plan.sub_steps.extend(
+        _generate_control_plane_plan(target, analysis_result.apps_control_plane, args.force)
     )
-    plan.add_step(control_plane_subordinate_upgrade_plan)
+    # plan.sub_steps.extend(
```
Comment (gabrielcocenza) [inline cou/steps/plan.py:434]: I guess you can remove this comment and the `# pragma: no cover`

File: cou/steps/plan.py [hunk 3]
```diff
@@ -317,6 +318,75 @@ def _get_os_release_and_series(analysis_result: Analysis) -> tuple[OpenStackRele
...
+        for app in apps
+        for unit in app.units.values()
+        if app.charm == "nova-compute"
+    ]
+    hypervisor_apps = []
+    for app in apps:
+        for machine in app.machines:
+            if machine in nova_compute_machines:
+                hypervisor_apps.append(app)
+                break  # exiting machine for loop, since
```
Comment (gabrielcocenza) [inline cou/steps/plan.py:380]: This is another reason why I think that the machine should have knowledge about if there is nova-compute on it or not. I think #248 should be merged first than this PR. With that I think this logic will be changed.

File: cou/steps/plan.py [hunk 4]
```diff
@@ -317,6 +318,75 @@ def _get_os_release_and_series(analysis_result: Analysis) -> tuple[OpenStackRele
...
+
+    hypervisor_planner = HypervisorUpgradePlanner(hypervisor_apps)
+    hypervisor_plans = hypervisor_planner.generate_upgrade_plan(target, force)
+    logger.debug("generating of hypervisors upgrade plan finished")
+
+    # generating rest
```
Comment (gabrielcocenza) [inline cou/steps/plan.py:386]: didn't understand this comment

File: cou/steps/plan.py (hunk 1 above)
Comment (rgildein) [reply cou/steps/plan.py:351]: I think you want to suggest to change first plan to plane.

File: cou/steps/plan.py (hunk 2 above)
Comment (rgildein) [reply cou/steps/plan.py:434]: I cannot, since I do not add any tests for this.

File: cou/steps/plan.py (hunk 3 above)
Comment (rgildein) [reply cou/steps/plan.py:380]: I agree.

File: cou/steps/plan.py (hunk 4 above)
Comment (rgildein) [reply cou/steps/plan.py:386]: For example, the ceph-osd will be there.

**Comment by gabrielcocenza** [comment]
/canonical/self-hosted-runners/run-workflows eadbc4d

**Comment by gabrielcocenza** [comment]
/canonical/self-hosted-runners/run-workflows eadbc4d1e45027e21389746e688d5255d3b5c1fc

**Comment by agileshaw** [comment]
/canonical/self-hosted-runners/run-workflows 99bfc316ab68d24684fececc6f3db83f988db884

File: cou/steps/plan.py [hunk 5]
```diff
@@ -293,6 +294,75 @@ def _get_os_release_and_series(analysis_result: Analysis) -> tuple[OpenStackRele
...
+        for app in apps
+        for unit in app.units.values()
+        if app.charm == "nova-compute"
+    ]
+    hypervisor_apps = []
+    for app in apps:
+        for machine in app.machines:
+            if machine in nova_compute_machines:
+                hypervisor_apps.append(app)
+                break  # exiting machine for loop, since
```
Comment (aieri) [inline cou/steps/plan.py:356]: truncated comment

File: cou/steps/plan.py [hunk 6]
```diff
@@ -293,6 +294,75 @@ def _get_os_release_and_series(analysis_result: Analysis) -> tuple[OpenStackRele
...
+
+    hypervisor_planner = HypervisorUpgradePlanner(hypervisor_apps)
+    hypervisor_plans = hypervisor_planner.generate_upgrade_plan(target, force)
+    logger.debug("generating of hypervisors upgrade plan finished")
```
Comment (aieri) [inline cou/steps/plan.py:360]:
```suggestion
    logger.debug("Generation of the hypervisors upgrade plan complete")
```

**Review by aieri — APPROVED** [review]
Verbiage nits, otherwise +1

File: cou/steps/plan.py (hunk 3 above)
Comment (gabrielcocenza) [reply cou/steps/plan.py:356]: We can let this to be changed later, so solving it.

File: cou/steps/plan.py [hunk 7]
```diff
@@ -293,6 +294,75 @@ def _get_os_release_and_series(analysis_result: Analysis) -> tuple[OpenStackRele
...
+        unit.machine.machine_id
+        for app in apps
+        for unit in app.units.values()
+        if app.charm == "nova-compute"
+    ]
+    hypervisor_apps = []
+    for app in apps:
+        for machine in app.machines:
+            if machine in nova_compute_machines:
+                hypervisor_apps.append(app)
```
Comment (gabrielcocenza) [inline cou/steps/plan.py:355]: I think we should keep ceph-osd out of the hypervisors_apps. This is also including subordinate charms. I guess it's not a issue now, but we need to keep an eye on it.

File: cou/steps/plan.py (hunk 4 above)
Comment (gabrielcocenza) [reply cou/steps/plan.py:362]:
```suggestion
    # generate the plan for the remaining data-plane apps
```
See my comment above of removing ceph-osd from hypervisor_apps.

File: cou/steps/hypervisor.py [hunk 8]
```diff
@@ -198,13 +198,15 @@ def _generate_post_upgrade_plan(self) -> PostUpgradeStep:
         raise NotImplementedError
 
     # pylint: disable=unused-argument
-    def generate_upgrade_plan(self, target: OpenStackRelease) -> UpgradePlan:
+    def generate_upgrade_plan(self, target: OpenStackRelease, force: bool) -> UpgradePlan:
```
Comment (gabrielcocenza) [inline cou/steps/hypervisor.py:201]: the planning so far is filtering the hypervisors machines. I think it makes sense accept this parameter in this function. If you think it should be part of another PR I'm fine with it, but we are already adding force so it might be an opportunity.

**Review by gabrielcocenza — COMMENTED** [review]
Just some comments to be discussed, but non-blockers

File: cou/steps/plan.py (hunk 4 above)
Comment (rgildein) [reply cou/steps/plan.py:362]: You know that this comment is just for you now and will be removed after we actually do ceph-osd.

File: cou/steps/hypervisor.py (hunk 8 above)
Comment (rgildein) [reply cou/steps/hypervisor.py:201]: Yes, and with that I need to add `__eq__` to COUMachie, since you cannot do `machine_id in machines`, where machines are `list[COUMachine]`.

Comment (gabrielcocenza) [reply cou/steps/hypervisor.py:201]: I'm assuming that won't be part of this PR, so solving it

**Review by gabrielcocenza — APPROVED** [review]
This PR has some conflicts with the one that I'm proposing #255 to separate control_plane from data_plane plan generation. However, I don't see a problem merging this and deal with the detail on this other PR

**Review by agileshaw — APPROVED** [review]
LGTM.
